tools/autograph-monitor/monitor.go

```diff
@@ -0,0 +1,361 @@
+}
+
+type configuration struct {
+	URL           string `yaml:"url"`
+	MonitoringKey string `yaml:"monitoringkey"`
+	RootHash      string `yaml:"security.content.signature.root_hash"`
+}
+
+var conf configuration
+
+const inputdata string = "AUTOGRAPH MONITORING"
+
+func main() {
+	var err error
+	confdir := "."
+	if os.Getenv("LAMBDA_TASK_ROOT") != "" {
+		confdir = os.Getenv("LAMBDA_TASK_ROOT")
+	}
+	// load the local configuration file
+	conf, err = loadConf(confdir + "/monitor.autograph.yaml")
+	if err != nil {
+		log.Fatalf("failed to load configuration: %v", err)
+	}
+
+	log.Println("Retrieving monitoring data from", conf.URL)
+	req, err := http.NewRequest("GET", conf.URL+"__monitor__", nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", makeAuthHeader(req, "monitor", conf.MonitoringKey))
+	cli := &http.Client{}
+	resp, err := cli.Do(req)
+	if err != nil || resp == nil {
+		log.Fatal(err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusCreated {
+		log.Fatalf("Request failed with %q", resp.Status)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// verify that we got a proper signature response, with a valid signature
+	var responses []signatureresponse
+	err = json.Unmarshal(body, &responses)
+	if err != nil {
+		log.Fatal(err)
+	}
+	failed := false
+	for i, response := range responses {
+		err = verify(response)
+		if err != nil {
+			failed = true
+			log.Printf("Response %d does not pass: %v", i, err)
+			log.Printf("Response was: %+v", response)
+		} else {
+			log.Printf("Response %d passes verification", i)
+		}
+	}
+	if failed {
+		log.Fatal("Errors found during monitoring")
+	}
+}
+
+// Verify the signature and certificate chain of a response.
+//
+// If an X5U value was provided, use the public key from the end entity certificate
+// to verify the sig. Otherwise, use the PublicKey contained in the response.
+//
+// If the signature passes, verify the chain of trust maps.
+func verify(sr signatureresponse) error {
+	var (
+		pubkey *ecdsa.PublicKey
+		err    error
+		data   []byte
+		certs  []*x509.Certificate
+	)
+	if sr.X5U != "" {
+		certs, err = getX5U(sr.X5U)
+		if err != nil {
+			return err
+		}
+		if len(certs) < 2 {
+			return fmt.Errorf("Found %d certs in X5U, expected at least 2", len(certs))
+		}
+		// certs[0] is the end entity
+		pubkey = certs[0].PublicKey.(*ecdsa.PublicKey)
+	} else {
+		pubkey, err = parsePublicKeyFromB64(sr.PublicKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	if sr.ContentSignature != "" {
+		data = make([]byte, len("Content-Signature:\x00")+len(inputdata))
+		copy(data[:len("Content-Signature:\x00")], []byte("Content-Signature:\x00"))
+		copy(data[len("Content-Signature:\x00"):], inputdata)
+	} else {
+		data = make([]byte, len(inputdata))
+		copy(data, inputdata)
+	}
+	datahash, err := digest(data, sr.Hash)
```

Comment:
This looks pretty complicated but I guess that's the way you have to do it in go.